loader/src/main/java/net/neoforged/fml/common/asm/enumextension/EnumParameters.java

```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.fml.common.asm.enumextension;
+
+import java.util.List;
+import org.objectweb.asm.Type;
+
+public interface EnumParameters {
```

Comment:
```suggestion
public sealed interface EnumParameters permits EnumParameters.Constant, EnumParameters.ListBased {
```
- Using sealed class here could optimize `RuntimeEnumExtender#loadConstructorParams` with enhanced switch.
- The subclasses names are a bit confusing, maybe `ListBacked` -> `FieldReference`?
- Supporting loading parameters from static methods should also be a choice.

Reply:
Implemented the first two suggestions.

How would static methods work here? One method per parameter (more type safe but annoying for more than one or two parameters)? Single method with an index parameter (lacking type-safety like the list but simple to implement)?

Reply:
It could also be expected to return a List, similar to the field, but potentially simpler to write.

Reply:
I would prefer to keep the generated bytecode close to how the compiler would generate it, mainly because it means that decompiling a transformed class for debugging works better. This means that I would have to call that method once per parameter which would in turn require the modder to store the list in a field and return that, which completely defeats the point of using a method.

Reply:
```java
public static Object getParameter(int index) {
    return switch (index) {
        case 0 -> someObj;
        //...
    }
}

//With generics
public static <T> T getParameter(int index, Class<T> type) {
    return switch (index) {
        case 0 -> type.cast(someObj);
        //...
    }
}
```
There's no need to return a list when using a static method, the above example should be sufficient enough.
Additionally, the method could be made generic and pass in a `Class<T>` parameter to enforce the cast being done in the modder's method first, so any `ClassCastException` will not be thrown from the enum class.